bin/propolis-server/src/lib/migrate/compat.rs

```diff
@@ -0,0 +1,791 @@
+    )]
+    Length { this: usize, other: usize },
+
+    #[error("collection key {0} present in self but not other")]
+    KeyAbsent(String),
+
+    #[error("component {0} incompatible")]
+    Component(String, #[source] ComponentIncompatibility),
+}
+
+impl<T: CompatComponent> CompatCollection for HashMap<String, T> {
+    fn is_compatible_with(
+        &self,
+        other: &Self,
+    ) -> Result<(), CollectionIncompatibility> {
+        if self.len() != other.len() {
+            return Err(CollectionIncompatibility::Length {
+                this: self.len(),
+                other: other.len(),
+            });
+        }
+
+        for (key, this_val) in self.iter() {
+            let other_val = other.get(key).ok_or_else(|| {
+                CollectionIncompatibility::KeyAbsent(key.clone())
+            })?;
+
+            this_val.is_compatible_with(other_val).map_err(|e| {
+                CollectionIncompatibility::Component(key.clone(), e)
+            })?;
+        }
+
+        Ok(())
+    }
+}
+
+impl spec::Spec {
+    fn is_board_compatible(
+        &self,
+        other: &Self,
+    ) -> Result<(), BoardIncompatibility> {
+        self.is_chipset_compatible(other)?;
+
+        let this = &self.board;
+        let other = &other.board;
+        if this.cpus != other.cpus {
+            Err(BoardIncompatibility::CpuCount {
+                this: this.cpus,
+                other: other.cpus,
+            })
+        } else if this.memory_mb != other.memory_mb {
+            Err(BoardIncompatibility::MemorySize {
+                this: this.memory_mb,
+                other: other.memory_mb,
+            })
+        } else {
+            Ok(())
+        }
+    }
+
+    fn is_chipset_compatible(
+        &self,
+        other: &Self,
+    ) -> Result<(), BoardIncompatibility> {
+        let Chipset::I440Fx(this) = self.board.chipset;
+        let Chipset::I440Fx(other) = other.board.chipset;
+
+        if this.enable_pcie != other.enable_pcie {
+            Err(BoardIncompatibility::PcieEnabled {
+                this: this.enable_pcie,
+                other: other.enable_pcie,
+            })
+        } else {
+            Ok(())
+        }
+    }
+
+    fn is_pvpanic_compatible(
+        &self,
+        other: &Self,
+    ) -> Result<(), PvpanicIncompatibility> {
+        match (&self.pvpanic, &other.pvpanic) {
+            (Some(this), Some(other)) => {
+                if this.name != other.name {
+                    Err(PvpanicIncompatibility::Name {
+                        this: this.name.clone(),
+                        other: other.name.clone(),
+                    })
+                } else if this.spec.enable_isa != other.spec.enable_isa {
+                    Err(PvpanicIncompatibility::EnableIsa {
+                        this: this.spec.enable_isa,
+                        other: other.spec.enable_isa,
+                    })
+                } else {
+                    Ok(())
+                }
+            }
```

Comment:
nit, take it or leave it: i probably would've roll the `if` guards into the `match`, as in 
```suggestion
            (Some(this), Some(other)) if this.name != other.name => {
                Err(PvpanicIncompatibility::Name {
                    this: this.name.clone(),
                    other: other.name.clone(),
                })
            },
            (Some(this), Some(other)) if this.spec.enable_isa != other.spec.enable_isa => {
                Err(PvpanicIncompatibility::EnableIsa {
                    this: this.spec.enable_isa,
                    other: other.spec.enable_isa,
                })
            },
            (Some(_), Some(_)) => Ok(()),
```
you could do this, if you like, but i'll freely admit it doesn't actually matter :)